oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -163,15 +169,74 @@ impl GrpcClientNode {
             }
         };
 
-        let body_stream = rsp_stream.get_mut();
+        let mut response_handler =
+            ResponseHandler::new(runtime.clone(), rsp_stream, invocation, method_name);
+        response_handler.handle().await
+    }
+
+    /// Creates a TLS connection to an external gRPC service.
+    async fn connect(
+        &self,
+    ) -> Result<tonic::client::Grpc<tonic::transport::channel::Channel>, tonic::transport::Error>
+    {
+        debug!("Connecting to {}", self.uri);
+
+        // Create a TLS configuration.
+        let tls_config = ClientTlsConfig::new().ca_certificate(self.root_tls_certificate.clone());
+
+        // Connect to a remote gRPC service.
+        let connection = Channel::builder(self.uri.clone())
+            .tls_config(tls_config)
+            .connect()
+            .await?;
+
+        debug!("Connected to {}", self.uri);
+        Ok(tonic::client::Grpc::new(connection))
+    }
+}
+
+struct ResponseHandler<'a> {
+    runtime: RuntimeProxy,
+    response_stream: tonic::Response<tonic::Streaming<Vec<u8>>>,
+    invocation: &'a Invocation,
+    method_name: String,
+    // The lifetime of the timer matches the lifetime of the iterator,
+    // updating the request-timer metric when the iterator is dropped.
+    _timer: prometheus::HistogramTimer,
+}
+
+impl<'a> ResponseHandler<'a> {
+    fn new(
+        runtime: RuntimeProxy,
+        response_stream: tonic::Response<tonic::Streaming<Vec<u8>>>,
+        invocation: &'a Invocation,
+        method_name: String,
+    ) -> Self {
+        ResponseHandler {
+            _timer: runtime
+                .metrics_data()
+                .grpc_client_metrics
+                .start_timer(&method_name),
+            runtime,
+            invocation,
+            response_stream,
+            method_name,
+        }
+    }
+
+    async fn handle(&mut self) -> Result<(), OakStatus> {
+        let body_stream = self.response_stream.get_mut();
         loop {
+            let invocation = self.invocation;
```

Comment:
nit: This local `invocation` is only used once &ndash; could maybe just use `self.location` at that spot.